src/otx/engine/adaptive_bs/adaptive_bs_api.py

```diff
@@ -0,0 +1,174 @@
+    engine: Engine,
+    not_increase: bool = True,
+    callbacks: list[Callback] | Callback | None = None,
+    **train_args,
+) -> None:
+    """Change the actual batch size depending on the current GPU status.
+
+    If not_increase is True, check current batch size is available to GPU and if not, decrease batch size.
+    If not_increase is False, increase batch size to use most of GPU memory.
+
+    Args:
+        engine (Engine): engine instnace.
+        not_increase (bool) : Whether adapting batch size to larger value than default value or not.
+        callbacks (list[Callback] | Callback | None, optional): callbacks used during training. Defaults to None.
+    """
+    if not (is_cuda_available() or is_xpu_available()):
+        msg = "Adaptive batch size supports CUDA or XPU."
+        raise RuntimeError(msg)
+    if engine.task == OTXTaskType.ZERO_SHOT_VISUAL_PROMPTING:  # type: ignore[has-type]
+        msg = "Zero shot visual prompting task doesn't support adaptive batch size."
+        raise RuntimeError(msg)
+
+    engine.model.patch_optimizer_and_scheduler_for_hpo()
+    default_bs = engine.datamodule.config.train_subset.batch_size
+
+    if "ADAPTIVE_BS_FOR_DIST" in os.environ:  # main process of distributed training already executes adapt_batch_size
+        new_batch_size = int(os.environ["ADAPTIVE_BS_FOR_DIST"])
+        if default_bs != new_batch_size:
+            _apply_new_batch_size(engine, new_batch_size)
+        return
+
+    train_func = partial(_train_model, engine=engine, callbacks=callbacks, **_adjust_train_args(train_args))
+    bs_search_algo = BsSearchAlgo(
+        train_func=train_func,
+        default_bs=default_bs,
+        max_bs=(
+            len(engine.datamodule.subsets[engine.datamodule.config.train_subset.subset_name]) // engine.device.devices
+        ),
+    )
+    if not_increase:
+        new_batch_size = bs_search_algo.auto_decrease_batch_size()
+    else:
+        new_batch_size = bs_search_algo.find_big_enough_batch_size()
+
+    if engine.device.devices != 1:
+        os.environ["ADAPTIVE_BS_FOR_DIST"] = str(new_batch_size)
+
+    if default_bs != new_batch_size:
+        origin_lr = engine.model.optimizer_callable.optimizer_kwargs["lr"]  # type: ignore[attr-defined]
+        _apply_new_batch_size(engine, new_batch_size)
+        msg = (
+            "Adapting batch size is done.\n"
+            f"Batch size is adapted : {default_bs} -> {new_batch_size}\n"
+            f"learning rate is adapted : {origin_lr} -> {engine.model.optimizer_callable.optimizer_kwargs['lr']}"  # type: ignore[attr-defined]
+        )
+        logger.info(msg)
+    else:
+        logger.info("Adapting batch size is done. Batch size isn't changed.")
+
+
+def _adjust_train_args(train_args: dict[str, Any]) -> dict[str, Any]:
+    train_args.update(train_args.pop("kwargs", {}))
+    train_args.pop("self", None)
+    train_args.pop("run_hpo", None)
+    train_args.pop("adaptive_bs")
+    return train_args
```

Comment:
Because of `**locals()` in `engine.train()`, this function seems required.
Can we use only necessary arguments instead of `**locals()`?

Reply:
it can, but IMHO, `train_args` is more readable because developer doesn't have to consider which arguments are only for `adapt_batch_size`. And also everytime argument of engine.train is changed, argument of `adapt_batch_size` also needs to be changed if it's changed. For these reasons, I think it's better to leave it as it is. How do you think about?

Reply:
I agree with your thought that users don't have to handle those args that can be dynamically changed.
But I'm just wondering if there is another way to except those args more cleanly.
Leave it for now but please consider it.